### ISSUE TITLE
[Java] Auto meta shared mode for type forward/backward compatibility 

### ISSUE THREAD

**Issue by chaokunyang** [apache/fory#202]
**Is your feature request related to a problem? Please describe.**
We supported type forward/backward compatibility in #197, but as the issue said, the solution proposed in that issue will write class meta every time the object is serialized. If multiple objects of same time are serialized as a whole, the meta will be serialized multiple times, which is unneccessary. 

We can use meta sharing to write meta only once in a serialization for an object graph. And the meta can be encoded to binary, so the actual meta writting will be just a memory copy, which is far more faster.

The issue #80 proposed meta share across serialization, which need the rpc or users to maintain the `MetaContext`, which is inconvinient for users. 

**Describe the solution you'd like**
We can  support auto meta sharing to reduce meta cost in every serialization. This will ensure multiple objects of same type write meta only once for space saving, and got better pperformance by memory copy meta binary.

Currently meta encoding is not compressed, this will be discussed in another issue.

**Additional context**
#80  #197
